### PR TITLE
Bug 2089093: CVO hotloops on OperatorGroup due to the diff of "upgradeStrategy"

### PR DESCRIPTION
**Pull request by Davoska**
This pull request will add `/apis/operators.coreos.com/v1/operatorgroups` handling that enables a custom way of processing manifests representing the `OperatorGroup` resource. Currently, the `OperatorGroup` resources are reconciled as generic resources. This doesn't provide defaulting of the fields.

This pull request will provide more precise processing of the `OperatorGroup` resources and will provide defaulting of the resources, which will fix the current hotlooping of the CVO.

Currently, the CVO hotloops on the `OperatorGroup` resources due to the missing defaulting of the resource (Bug 2089093 [1]).

The [2] commit introduced the `UpgradeStrategy` field to the `OperatorGroup` type. This field has a default value of `"Default"`. The CVO currently doesn't know of this default value, and thus the CVO hotloops when handling the `OperatorGroup` resource as a generic resource because some manifests don't specify the value of the field and rely on defaulting [3].

This pull request will provide a more precise way of processing the `OperatorGroup` type and will also fix the mentioned bug by providing the defaulting for the resource.

[1] https://bugzilla.redhat.com/show_bug.cgi?id=2089093
[2] openshift/operator-framework-olm@d895574
[3] https://github.com/openshift/operator-framework-olm/blob/ab5ecfd7f235142f191d91d103afabccdfd6ff04/manifests/0000_50_olm_13-operatorgroup-default.yaml

### CODE REVIEW

**Comment by Davoska** [comment]
In the [artifacts/**e2e-agnostic-upgrade-out-of-change**/gather-extra/artifacts/pods/openshift-cluster-version_cluster-version-operator-866d576bdc-mcbl6_cluster-version-operator.log](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/862/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic-upgrade-out-of-change/1590401075574214656/artifacts/e2e-agnostic-upgrade-out-of-change/gather-extra/artifacts/pods/openshift-cluster-version_cluster-version-operator-866d576bdc-mcbl6_cluster-version-operator.log) is not present any hotlooping of the CVO on the `OperatorGroup` while **in** the change. After upgrading to **out** of change the hotlooping starts as expected.

---

Checking for hotlooping in [artifacts/**e2e-agnostic**/gather-extra/artifacts/pods/openshift-cluster-version_cluster-version-operator-799578f87f-c77rk_cluster-version-operator.log](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/862/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic/1590401075406442496/artifacts/e2e-agnostic/gather-extra/artifacts/pods/openshift-cluster-version_cluster-version-operator-799578f87f-c77rk_cluster-version-operator.log) results in:

```bash
$ grep -o 'Updating .*due to diff' openshift-cluster-version_cluster-version-operator-799578f87f-c77rk_cluster-version-operator.log | sort | uniq -c
     14 Updating CRD alertmanagerconfigs.monitoring.coreos.com due to diff
     14 Updating CRD consoleplugins.console.openshift.io due to diff
     14 Updating CRD performanceprofiles.performance.openshift.io due to diff
     14 Updating CronJob openshift-operator-lifecycle-manager/collect-profiles due to diff
     14 Updating ValidatingWebhookConfiguration /controlplanemachineset.machine.openshift.io due to diff
     14 Updating ValidatingWebhookConfiguration /performance-addon-operator due to diff
```

There no longer is any hotlooping on the `OperatorGroup` present.

---

Checking for hotlooping in the `4.12.0-0.ci-2022-11-09-070828` release resulted in:

```bash
$ grep -o 'Updating .*due to diff' cluster-version-operator-78d75776c-dgxjd-cluster-version-operator.log | sort | uniq -c
     20 Updating CRD alertmanagerconfigs.monitoring.coreos.com due to diff
     20 Updating CRD consoleplugins.console.openshift.io due to diff
     20 Updating CRD performanceprofiles.performance.openshift.io due to diff
     20 Updating CronJob openshift-operator-lifecycle-manager/collect-profiles due to diff
     20 Updating OperatorGroup openshift-monitoring/openshift-cluster-monitoring due to diff
     20 Updating OperatorGroup openshift-operator-lifecycle-manager/olm-operators due to diff
     20 Updating OperatorGroup openshift-operators/global-operators due to diff
     20 Updating ValidatingWebhookConfiguration /controlplanemachineset.machine.openshift.io due to diff
     20 Updating ValidatingWebhookConfiguration /performance-addon-operator due to diff
```

As we can see after comparing the output to the previous output, no new hotlooping is present and only the `OperatorGroup` hotlooping disappeared.

---

Checking for any mentions of `OperatorGroup`  in [artifacts/**e2e-agnostic**/gather-extra/artifacts/pods/openshift-cluster-version_cluster-version-operator-799578f87f-c77rk_cluster-version-operator.log](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/862/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic/1590401075406442496/artifacts/e2e-agnostic/gather-extra/artifacts/pods/openshift-cluster-version_cluster-version-operator-799578f87f-c77rk_cluster-version-operator.log) returns following:

```bash
grep -io '].*operatorgroup.*' openshift-cluster-version_cluster-version-operator-799578f87f-c77rk_cluster-version-operator.log | sort | uniq -c
     27 ] Done syncing for customresourcedefinition "operatorgroups.operators.coreos.com" (608 of 828)
     28 ] Done syncing for operatorgroup "openshift-monitoring/openshift-cluster-monitoring" (728 of 828)
     26 ] Done syncing for operatorgroup "openshift-operator-lifecycle-manager/olm-operators" (632 of 828)
     26 ] Done syncing for operatorgroup "openshift-operators/global-operators" (631 of 828)
      1 ] error running apply for operatorgroup "openshift-monitoring/openshift-cluster-monitoring" (728 of 828): operatorgroups.operators.coreos.com "openshift-cluster-monitoring" is forbidden: User "system:serviceaccount:openshift-cluster-version:default" cannot get resource "operatorgroups" in API group "operators.coreos.com" in the namespace "openshift-monitoring"
     27 ] Running sync for customresourcedefinition "operatorgroups.operators.coreos.com" (608 of 828)
     28 ] Running sync for operatorgroup "openshift-monitoring/openshift-cluster-monitoring" (728 of 828)
     26 ] Running sync for operatorgroup "openshift-operator-lifecycle-manager/olm-operators" (632 of 828)
     26 ] Running sync for operatorgroup "openshift-operators/global-operators" (631 of 828)
```

Syncing seems to be working properly. The one error seems to be irrelevant to the PR as the same error happened to multiple resources at the time and then fixed itself.

File: lib/resourcemerge/operators_test.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			defaultOperatorGroup(&test.existing, test.existing)
+			defaultOperatorGroup(&test.expected, test.expected)
+			modified := pointer.BoolPtr(false)
```
Comment (petr-muller) [inline lib/resourcemerge/operators_test.go:218]:
```suggestion
			modified := pointer.Bool(false)
```

File: lib/resourcemerge/operators_test.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+			if !equality.Semantic.DeepEqual(test.existing, test.expected) {
+				t.Errorf("mismatch OperatorGroup got: %v want: %v", test.existing, test.expected)
+			}
+		})
+	}
+}
+
+// Ensures the structure contains any defaults not explicitly set by the test
+func defaultOperatorGroup(in *operatorsv1.OperatorGroup, from operatorsv1.OperatorGroup) {
+	modified := pointer.BoolPtr(false)
```
Comment (petr-muller) [inline lib/resourcemerge/operators_test.go:233]:
```suggestion
	modified := pointer.Bool(false)
```

**Review by petr-muller — COMMENTED** [review]
LGTM with nits, suggestions inline. 

As a newbie here, it seems a little weird to me that we need special treatment code in CVO for each different resource type, and always get it right, otherwise we get these hotloops/not reconciling bugs... I wonder if there is a better way. But that is of course out of scope of this PR - this change just adds a treatment for `OperatorGroup` that other types have.

**Review by petr-muller — APPROVED** [review]

**Comment by petr-muller** [comment]
Hmm GH approval != `/lgtm` here? :thinking: 

/lgtm

File: lib/resourceapply/operators.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	}
+	// if we only create this resource, we have no need to continue further
+	if IsCreateOnly(required) {
+		return nil, false, nil
+	}
+
+	var original operatorsv1.OperatorGroup
+	existing.DeepCopyInto(&original)
+
+	modified := pointer.BoolPtr(false)
```
[the remark this answers is not in the document]
Comment (Davoska) [reply lib/resourceapply/operators.go]: Thanks, Petr! I have forgotten about the https://github.com/openshift/cluster-version-operator/commit/f584cafba942440ef372a1026270fc47a4dc8a4d commit cleaning some of the deprecated function calls.

Fixed in the new commit https://github.com/openshift/cluster-version-operator/commit/2f0078a35636946231d5153645061744489cb3fc.

File: lib/resourcemerge/operators_test.go (hunk 1 above)
Comment (Davoska) [reply lib/resourcemerge/operators_test.go]: Fixed in the new commit https://github.com/openshift/cluster-version-operator/commit/2f0078a35636946231d5153645061744489cb3fc.

**Review by petr-muller — APPROVED** [review]

**Comment by Davoska** [comment]
As for https://github.com/openshift/cluster-version-operator/pull/862#pullrequestreview-1177370699:
> ... it seems a little weird to me that we need special treatment code in CVO for each different resource type, and always get it right, otherwise we get these hotloops/not reconciling bugs...

I agree that the current code (_the special treatment for each different resource type_) isn't great for making these kinds of new changes. The chance of shooting yourself in the foot is quite big when making changes. Hopefully, in the future, we could try to address this issue. However, this type of code doesn't need to be changed frequently.

**Comment by petr-muller** [comment]
> I agree that the current code (the special treatment for each different resource type) isn't great for making these kinds of new changes. The chance of shooting yourself in the foot is quite big when making changes. Hopefully, in the future, we could try to address this issue. However, this type of code doesn't need to be changed frequently.

I guess SSA could be a (partial?) solution: https://kubernetes.io/blog/2022/10/20/advanced-server-side-apply/

**Review by petr-muller — APPROVED** [review]
LGTM

**Pull request merged**